File: src/Cxx/PolyData/Curvatures.cxx
#include <vtkSmartPointer.h>

#include <vtkCurvatures.h>

#include <vtkXMLPolyDataReader.h>

#include <vtkLookupTable.h>
#include <vtkColorTransferFunction.h>
#include <vtkColorSeries.h>

#include <vtkPointData.h>

#include <vtkPolyDataMapper.h>
#include <vtkActor.h>
#include <vtkScalarBarActor.h>
#include <vtkRenderWindow.h>
#include <vtkRenderer.h>
#include <vtkRenderWindowInteractor.h>

int main(int argc, char *argv[])
{
  // Parse command line arguments
  if(argc < 2)
  {
    std::cerr << "Usage: " << argv[0] << " Filename(.vtp)" << std::endl;
    return EXIT_FAILURE;
  }

  // Create a polydata
  vtkSmartPointer<vtkXMLPolyDataReader> reader =
    vtkSmartPointer<vtkXMLPolyDataReader>::New();
  reader->SetFileName(argv[1]);

  vtkSmartPointer<vtkCurvatures> curvaturesFilter =
    vtkSmartPointer<vtkCurvatures>::New();
  curvaturesFilter->SetInputConnection(reader->GetOutputPort());
  curvaturesFilter->SetCurvatureTypeToMinimum();
  curvaturesFilter->SetCurvatureTypeToMaximum();
  curvaturesFilter->SetCurvatureTypeToGaussian();
  curvaturesFilter->SetCurvatureTypeToMean();
  curvaturesFilter->Update();

  // Get scalar range from command line if present, otherwise use
  // range of computed curvature
  double scalarRange[2];
  if (argc >= 4)
  {
    scalarRange[0] = atof(argv[2]);
    scalarRange[1] = atof(argv[3]);
  }
  else
  {
    curvaturesFilter->GetOutput()->GetScalarRange(scalarRange);
  }

  int scheme = 16;
  if (argc >= 5)
  {
    scheme = atoi(argv[4]);
  }

  // Build a lookup table
  vtkSmartPointer<vtkColorSeries> colorSeries =
    vtkSmartPointer<vtkColorSeries>::New();
  colorSeries->SetColorScheme(scheme);
  std::cout << "Using color scheme #: "
            << colorSeries->GetColorScheme() << " is "
            << colorSeries->GetColorSchemeName() << std::endl;

  vtkSmartPointer<vtkColorTransferFunction> lut =
    vtkSmartPointer<vtkColorTransferFunction>::New();
  lut->SetColorSpaceToHSV();

  // Use a color series to create a transfer function
  int numColors = colorSeries->GetNumberOfColors();
  for (int i = 0; i < numColors; i++)
  {
      vtkColor3ub color = colorSeries->GetColor(i);
    double dColor[3];
    dColor[0] = static_cast<double> (color[0]) / 255.0;
    dColor[1] = static_cast<double> (color[1]) / 255.0;
    dColor[2] = static_cast<double> (color[2]) / 255.0;
    double t = scalarRange[0] + (scalarRange[1] - scalarRange[0])
      / (numColors - 1) * i;
    lut->AddRGBPoint(t, dColor[0], dColor[1], dColor[2]);
  }

  // Create a mapper and actor
  vtkSmartPointer<vtkPolyDataMapper> mapper =
    vtkSmartPointer<vtkPolyDataMapper>::New();
  mapper->SetInputConnection(curvaturesFilter->GetOutputPort());
  mapper->SetLookupTable(lut);
  mapper->SetScalarRange(scalarRange);

  vtkSmartPointer<vtkActor> actor =
    vtkSmartPointer<vtkActor>::New();
  actor->SetMapper(mapper);

  // Create a scalar bar
  vtkSmartPointer<vtkScalarBarActor> scalarBar =
    vtkSmartPointer<vtkScalarBarActor>::New();
  scalarBar->SetLookupTable(mapper->GetLookupTable());
  scalarBar->SetTitle(
    curvaturesFilter->GetOutput()->GetPointData()->GetScalars()->GetName());
  scalarBar->SetNumberOfLabels(5);

  // Create a renderer, render window, and interactor
  vtkSmartPointer<vtkRenderer> renderer =
    vtkSmartPointer<vtkRenderer>::New();
  vtkSmartPointer<vtkRenderWindow> renderWindow =
    vtkSmartPointer<vtkRenderWindow>::New();
  renderWindow->AddRenderer(renderer);
  vtkSmartPointer<vtkRenderWindowInteractor> renderWindowInteractor =
    vtkSmartPointer<vtkRenderWindowInteractor>::New();
  renderWindowInteractor->SetRenderWindow(renderWindow);

  // Add the actors to the scene
  renderer->AddActor(actor);
  renderer->AddActor2D(scalarBar);

  renderer->SetBackground(.1, .2, .3); // Background color blue

  // Render and interact
  renderWindow->Render();
  renderWindowInteractor->Start();

  return EXIT_SUCCESS;
}

File: src/Python/PolyData/Curvatures.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
from __future__ import print_function, division
import vtk

def main(filename, curvature=0, scalarRange=None, scheme=None):
    print("Loading", filename)
    reader = vtk.vtkXMLPolyDataReader()
    reader.SetFileName(filename)

    curvaturesFilter = vtk.vtkCurvatures()
    curvaturesFilter.SetInputConnection(reader.GetOutputPort())
    if curvature == 0:
        curvaturesFilter.SetCurvatureTypeToMinimum()
    elif curvature == 1:
        curvaturesFilter.SetCurvatureTypeToMaximum()
    elif curvature == 2:
        curvaturesFilter.SetCurvatureTypeToGaussian()
    else:
        curvaturesFilter.SetCurvatureTypeToMean()
    curvaturesFilter.Update()

    # Get scalar range from command line if present, otherwise use
    # range of computed curvature
    if scalarRange is None:
        scalarRange = curvaturesFilter.GetOutput().GetScalarRange()

    # Build a lookup table
    if scheme is None:
        scheme = 16
    colorSeries = vtk.vtkColorSeries()
    colorSeries.SetColorScheme(scheme)
    print("Using color scheme #:", colorSeries.GetColorScheme(), \
        "is", colorSeries.GetColorSchemeName())

    lut = vtk.vtkColorTransferFunction()
    lut.SetColorSpaceToHSV()

    # Use a color series to create a transfer function
    numColors = colorSeries.GetNumberOfColors()
    for i in range(numColors):
        color = colorSeries.GetColor(i)
        dColor = [color[0]/255.0, color[1]/255.0, color[2]/255.0]
        t = scalarRange[0] + (scalarRange[1] - scalarRange[0]) / (numColors - 1) * i
        lut.AddRGBPoint(t, dColor[0], dColor[1], dColor[2])

    # Create a mapper and actor
    mapper = vtk.vtkPolyDataMapper()
    mapper.SetInputConnection(curvaturesFilter.GetOutputPort())
    mapper.SetLookupTable(lut)
    mapper.SetScalarRange(scalarRange)

    actor = vtk.vtkActor()
    actor.SetMapper(mapper)

    # Create a scalar bar
    print("Displaying", curvaturesFilter.GetOutput().GetPointData().GetScalars().GetName())
    scalarBarActor = vtk.vtkScalarBarActor()
    scalarBarActor.SetLookupTable(mapper.GetLookupTable())
    scalarBarActor.SetTitle(
        curvaturesFilter.GetOutput().GetPointData().GetScalars().GetName())
    
    scalarBarActor.SetNumberOfLabels(5)

    # Create a renderer, render window, and interactor
    renderer = vtk.vtkRenderer()
    renderWindow = vtk.vtkRenderWindow()
    renderWindow.AddRenderer(renderer)
    renderWindowInteractor = vtk.vtkRenderWindowInteractor()
    renderWindowInteractor.SetRenderWindow(renderWindow)

    # Add the actors to the scene
    renderer.AddActor(actor)
    renderer.AddActor2D(scalarBarActor)

    renderer.SetBackground(.1, .2, .3) # Background color blue

    # Render and interact
    renderWindow.Render()
    renderWindowInteractor.Start()


def get_program_parameters():
    import argparse
    description = 'Computes the curvature of a polydata surface.'
    epilogue = '''
    filename=./src/Testing/Data/cowHead.vtp
    curvature: 0=Min, 1=Max, 2=Gauss, 3=Mean
    '''
    parser = argparse.ArgumentParser(description=description, epilog=epilogue,
                                     formatter_class=argparse.RawDescriptionHelpFormatter)
    parser.add_argument('filename', help='Filename.vtp')
    parser.add_argument('curvature', type=int, help='Int value')
    parser.add_argument('scalarRangeLow', nargs='?', type=float, help='Float value')
    parser.add_argument('scalarRangeHigh', nargs='?', type=float, help='Float value')
    parser.add_argument('colorScheme', nargs='?', type=int, help='Int value')
    args = parser.parse_args()
    scalarRange = None
    if args.scalarRangeLow and args.scalarRangeHigh:
        scalarRange = (args.scalarRangeLow, args.scalarRangeHigh)
    return args.filename, args.curvature, scalarRange, args.colorScheme

if __name__ == "__main__":
    main(*get_program_parameters())
